### PR TITLE
[ec2_ami_facts] new boto3-based module as a replacement for ec2_ami_find

### PR DESCRIPTION
**Pull request by prasadkatti**
##### SUMMARY

- new boto3-based module to gather facts about ec2 images
- intended to replace ec2_ami_find which uses boto
- an ami find task (using new module) added to the ec2_ami integration test

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
`ec2_ami_facts`

##### ANSIBLE VERSION

```
ansible 2.5.0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_ami_facts.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    virtualization_type:
+      description: The type of virtualization of the AMI
+      returned: always
+      type: string
+      sample: hvm
+'''
+
+import traceback
+
+try:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_ami_facts.py]: No need to import boto3, and if you use AnsibleAWSModule, no need to import or test HAS_BOTO3.

File: lib/ansible/modules/cloud/amazon/ec2_ami_facts.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+    image_ids = module.params.get("image_ids")
+    filters = ansible_dict_to_boto3_filter_list(module.params.get("filters"))
+    owners = module.params.get("owners")
+    executable_users = module.params.get("executable_users")
+
+    try:
+        images = ec2_client.describe_images(ImageIds=image_ids, Filters=filters, Owners=owners, ExecutableUsers=executable_users)
+    except ClientError as err:
+        module.fail_json(msg=err.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(err.response))
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_ami_facts.py]: should catch BotoCoreError here too.

Using AnsibleAWSModule means you can just do module.fail_json_aws(e, msg="Couldn't list images")

File: test/integration/targets/ec2_ami/tasks/main.yml [hunk 3]
```diff
@@ -118,6 +118,23 @@
 
     # ============================================================
 
+    - name: gather facts about the image created
+      ec2_ami_facts:
```
Comment (willthames) [inline test/integration/targets/ec2_ami/tasks/main.yml:122]: Probably worth adding ec2_ami_facts to test/integration/targets/ec2_ami/aliases

**Review by willthames — CHANGES_REQUESTED** [review]
Might be worth deprecating ec2_ami_find at the same time: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_checklist.html#deprecating-and-making-module-aliases

Edit: I remember this caused chaos on a PR last time I suggested this. Deprecating ec2_ami_find should be a separate PR once this one is accepted, sorry.

File: test/integration/targets/ec2_ami/tasks/main.yml (hunk 3 above)
Comment (prasadkatti) [reply test/integration/targets/ec2_ami/tasks/main.yml:122]: I am not sure how this works. Does "ec2_ami_facts" need to be on a separate line by itself?

What effect does it have? Would that make the `ec2_ami` target run on any commits to the `ec2_ami_facts` module?

**Comment by prasadkatti** [comment]
ready_for_review

File: test/integration/targets/ec2_ami/tasks/main.yml (hunk 3 above)
Comment (willthames) [reply test/integration/targets/ec2_ami/tasks/main.yml:122]: You're quite right on both points.

File: lib/ansible/modules/cloud/amazon/ec2_ami_facts.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      description: The location of the AMI
+      returned: always
+      type: string
+      sample: 408466080000/Webapp
+    image_type:
+      description: The type of image
+      returned: always
+      type: string
+      sample: machine
+    name:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_ami_facts.py]: It's worth returning the launch permissions (you need to run `describe_image_attribute` separately to get this)

Comment (prasadkatti) [reply lib/ansible/modules/cloud/amazon/ec2_ami_facts.py]: @willthames - Thanks. I have made the requested changes.

**Comment by prasadkatti** [comment]
ready_for_review

**Review by willthames — APPROVED** [review]

**Review by Sodki — APPROVED** [review]

**Comment by Sodki** [comment]
shipit

**Comment by willthames** [comment]
Merged, thanks @prasadkatti. If you'd like to do the deprecation work on ec2_ami_find, including an update to the CHANGELOG, that would be well received too :)

**Comment by prasadkatti** [comment]
@willthames - Done. Here is the PR - #32501

**Pull request merged**